File: 백준/Gold/2573. 빙산/빙산.java


import java.util.*;
import java.io.*;


public class Main {

    static int N, M;
    static int[][] graph;
    static boolean[][] visited;


    static int[] dx = {0, 0, 1, -1};
    static int[] dy = {1, -1, 0, 0};

    static int time = 0;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

//        visited = new boolean[N][M];
        graph = new int[N][M];
        for(int i = 0; i<N; i++){
            st = new StringTokenizer(br.readLine());
            for(int j =0; j<M; j++){
                graph[i][j] = Integer.parseInt(st.nextToken());
            }
        }




        while(true){
            time++;

            graph = melt();

            int area = 0;
            visited = new boolean[N][M];

            for(int i = 0; i<N; i++){
                for(int j=0; j<M; j++){
                    if(graph[i][j] > 0 && !visited[i][j]){
                        bfs(i, j);
                        area++;
                    }
                }
            }


            if(area >= 2 ){
                System.out.println(time);
                break;
            }


            if(area ==0){
                System.out.println(0);
                break;
            }
        }




    }


    static int[][] melt(){
        int[][] temp = new int[N][M];

        for(int i = 0; i<N; i++){
            for(int j = 0; j<M; j++){
                int cnt = 0;

                if(graph[i][j] > 0){
                    for(int k = 0; k<4; k++){
                        int nx = i + dx[k];
                        int ny = j + dy[k];

                        if(0<=nx && nx <N && 0<= ny && ny<M && graph[nx][ny] == 0){
                            cnt++;
                        }
                    }
                }

                temp[i][j] = Math.max(graph[i][j] - cnt, 0);
            }
        }

        return temp;
    }


    static void bfs(int startX, int startY){
        Deque<Pos> q = new ArrayDeque<>();
        q.offer(new Pos(startX, startY));
        visited[startX][startY] = true;


        while(!q.isEmpty()){
            Pos p = q.poll();
            int x = p.x;
            int y = p.y;

            for(int i = 0; i<4; i++){
                int nx = x + dx[i];
                int ny = y + dy[i];

                if(0<= nx && nx<N && 0<=ny && ny<M && !visited[nx][ny] && graph[nx][ny] > 0){
                    q.offer(new Pos(nx, ny));
                    visited[nx][ny] = true;
                }

            }


        }




    }


    static class Pos{
        int x;
        int y;

        Pos(int x, int y){
            this.x = x;
            this.y = y;
        }
    }

}

File: 백준/Gold/2573. 빙산/빙산.py


import sys
input = sys.stdin.readline

from collections import deque


n, m = map(int, input().split(" "))
graph = []
visited = [[0 for _ in range(m)] for _ in range(n)]

dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]

time = 0


for _ in range(n):
    temp = list(map(int, input().split(" ")))
    graph.append(temp)


def bfs(start_x, start_y):
    queue = deque()

    queue.append((start_x, start_y))
    visited[start_x][start_y] = 1

    while queue:
        x, y = queue.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0<= nx and nx <n and 0<= ny and ny <m and not visited[nx][ny] and graph[nx][ny]:
                queue.append((nx, ny))
                visited[nx][ny] = 1



while True:
    time += 1

    # 녹기
    new_graph = [[0 for _ in range(m)] for _ in range(n)]

    for i in range(n):
        for j in range(m):
            water = 0

            if graph[i][j]:
                for nx, ny in ((i, j+1), (i, j-1), (i+1, j), (i-1, j)):
                    if 0<= nx and nx <n and 0<= ny and ny <m and graph[nx][ny] == 0:
                        water += 1

            new_graph[i][j] = max(0, graph[i][j] - water)
    graph = new_graph


    # 영역 개수 세기
    visited = [[0 for _ in range(m)] for _ in range(n)]
    cnt = 0

    for i in range(n):
        for j in range(m):
            if graph[i][j] and not visited[i][j]:
                bfs(i, j)
                cnt += 1

    if cnt >= 2:
        print(time)
        exit()


    if cnt == 0:
        print(0)
        exit()